crates/mcp-apollo-server/src/explorer.rs

```diff
@@ -43,17 +47,42 @@ impl Explorer {
     }
 
     fn create_explorer_url(&self, input: &Value) -> String {
-        let compressed = lz_str::compress_to_uint8_array(input.to_string().as_str());
-        let encoded = base64::engine::general_purpose::STANDARD.encode(compressed);
+        let mut input = input.clone();
+
+        let document = input.get("document").and_then(|v| v.as_str());
+        if document.is_none() || document == Some("") {
+            if let Some(obj) = input.as_object_mut() {
+                obj.insert("document".to_string(), Value::String("{}".to_string()));
+            }
+        }
+        let variables = input.get("variables").and_then(|v| v.as_str());
+        if variables.is_none() || variables == Some("") {
+            if let Some(obj) = input.as_object_mut() {
+                obj.insert("variables".to_string(), Value::String("{}".to_string()));
+            }
+        }
+        let headers = input.get("headers").and_then(|v| v.as_str());
+        if headers.is_none() || headers == Some("") {
+            if let Some(obj) = input.as_object_mut() {
+                obj.insert("headers".to_string(), Value::String("{}".to_string()));
+            }
+        }
+        let compressed = lz_str::compress_to_encoded_uri_component(input.to_string().as_str());
         format!(
-            "https://studio.apollographql.com/graph/{graph_id}/variant/{variant}/explorer?explorerURLState={encoded}",
+            "https://studio.apollographql.com/graph/{graph_id}/variant/{variant}/explorer?explorerURLState={compressed}",
             graph_id = self.graph_id,
             variant = self.variant
         )
     }
 
     pub async fn execute(&self, input: Value) -> Result<CallToolResult, McpError> {
-        webbrowser::open(self.create_explorer_url(&input).as_str())
+        let url = self.create_explorer_url(&input);
+        info!(
+            "Opening Apollo Explorer URL: {} for input operation: {}",
+            url,
+            serde_json::to_string_pretty(&input).unwrap_or("<unable to serialize>".into())
+        );
```

Comment:
`serde_json::to_string_pretty`already alocates, so if you shuffle these around you can avoid a clone.

```suggestion
        let pretty = serde_json::to_string_pretty(&input).unwrap_or_else(|| "<unable to serialize>".into());
        let url = self.create_explorer_url(input);
        info!(
            "Opening Apollo Explorer URL: {url} for input operation: {pretty}",
        );
```